Purchases/Networking/Operations/NetworkOperation.swift

```diff
@@ -45,17 +81,54 @@ class NetworkOperation: Operation {
     }
 
     override final func main() {
-        if isCancelled == true {
+        if self.isCancelled {
+            self.isFinished = true
             return
         }
 
-        begin()
+        self.isExecuting = true
+
+        Self.log("Started")
+        self.begin()
     }
 
+    override final func cancel() {
+        self.isCancelled = true
+        self.isExecuting = false
+        self.isFinished = true
+
+        Self.log("Cancelled")
+    }
+
+    /// Called by subclasses to complete this operation
+    final func finish() {
+        assert(!self.isFinished, "Operation \(type(of: self)) (\(self)) was already finished")
```

Comment:
Having to manually call `finish` is very error prone (but that's how async `Operations` work). Luckily this catches duplicate `finish` calls in tests.
Should we change it to `precondition` though?

Reply:
`precondition` will cause a crash if `finish` is called twice, right? If that is the case, that seems a big dangerous 🤔 

Can we early return after the the `assert` if its already finished?

Reply:
> `precondition` will cause a crash if finish is called twice, right? If that is the case, that seems a big dangerous 🤔

True, but sometimes "dangerous" might be better than data-loss or unexpected behavior, it's never an obvious answer.

> Can we early return after the the assert if its already finished?

Yeah but I think that wouldn't really change anything?

Reply:
This is a tough one. 
I don't love crashing in the SDK if we can avoid it, but it's nuanced. 
If for example, the problem would happen if an external service (our backend, storekit) behaves in an unexpected way, then I think we should never crash, and instead provide as seamless of an experience as we can. 

If on the other hand, the problem would happen strictly from our own programming bug, and we can confidently prevent that bug through automated tests, then that seems like a case where we can leverage the crash to prevent a bug from existing. 

It seems like this case falls under the latter category, so I'm all for `precondition`, but we should make sure that we're covering all instances where it could get called twice.

Reply:
thinking about this more, and the `defer` thing, where we need to make sure we call it on all `completion` blocks individually, perhaps we could improve this with a bit of a refactor? 

Like, what if `begin()` took in a completion block, so that in `NetworkOperation`'s `main`, we do something like:
```swift
    override final func main() {
        if isCancelled == true {
            return
        }

        begin { 
            self.isCompleted = true
        }
    }
```

Reply:
> If on the other hand, the problem would happen strictly from our own programming bug, and we can confidently prevent that bug through automated tests, then that seems like a case where we can leverage the crash to prevent a bug from existing.

That's why I went with `assert`, since our tests will catch that when running on DEBUG.

> thinking about this more, and the defer thing, where we need to make sure we call it on all completion blocks individually, perhaps we could improve this with a bit of a refactor?

Actually I think that's a great idea, let me try that. At the very least it makes it explicit in all subclasses that this completion block is needed, and it removes the duplication.

Reply:
Done!